app/org/sagebionetworks/bridge/play/controllers/BaseController.java

```diff
@@ -164,16 +167,27 @@ UserSession getAuthenticatedSession(boolean consentRequired, Roles...roles) {
         if (session == null || !session.isAuthenticated()) {
             throw new NotAuthenticatedException();
         }
+
+        // Sessions are locked to an IP address if (a) it is enabled in the study for unprivileged participant accounts
+        // or (b) always for privileged accounts.
+        Study study = studyService.getStudy(session.getStudyIdentifier());
+        Set<Roles> userRoles = session.getParticipant().getRoles();
+        boolean userHasRoles = !userRoles.isEmpty();
+        if (study.isParticipantIpLockingEnabled() || userHasRoles) {
+            String sessionIpAddress = parseIpAddress(session.getIpAddress());
+            String requestIpAddress = parseIpAddress(getRemoteAddress());
+            if (!Objects.equals(sessionIpAddress, requestIpAddress)) {
+                throw new NotAuthenticatedException();
+            }
+        }
```

Comment:
What if we tested, and only recorded the user's IP address if we needed to check it (enabled for entire study or has roles)? Then if the session had an IP address, we'd check against the request (or otherwise skip it). Then the study loading would only occur once and we probably already have it.